subsys/bluetooth/services/cgms/cgms_socp.c

```diff
@@ -0,0 +1,148 @@
+#define SOCP_SET_RATE_OF_INCREASE_ALERT_LEVEL         0x16
+#define SOCP_GET_RATE_OF_INCREASE_ALERT_LEVEL         0x17
+#define SOCP_RATE_OF_INCREASE_ALERT_LEVEL_RESPONSE    0x18
+#define SOCP_RESET_DEVICE_SPECIFIC_ALERT              0x19
+#define SOCP_START_THE_SESSION                        0x1A
+#define SOCP_STOP_THE_SESSION                         0x1B
+#define SOCP_RESPONSE_CODE                            0x1C
+
+#define SOCP_RSP_RESERVED_FOR_FUTURE_USE              0x00
+#define SOCP_RSP_SUCCESS                              0x01
+#define SOCP_RSP_OP_CODE_NOT_SUPPORTED                0x02
+#define SOCP_RSP_INVALID_OPERAND                      0x03
+#define SOCP_RSP_PROCEDURE_NOT_COMPLETED              0x04
+#define SOCP_RSP_OUT_OF_RANGE                         0x05
+
+extern uint16_t comm_interval;
+
+extern struct k_work_delayable report_meas_work;
+
+static int generic_handler(struct bt_conn *peer, uint8_t req_opcode, uint8_t response_code)
+{
+	uint8_t rsp[3];
+
+	rsp[0] = SOCP_RESPONSE_CODE;
+	rsp[1] = req_opcode;
+	rsp[2] = response_code;
+
+	int rc;
+
+	rc = cgms_socp_send_response(peer, rsp, sizeof(rsp));
+	return rc;
+}
+
+static int write_cgm_communication_interval_handler(struct bt_conn *peer,
+			const uint8_t *req_operand, uint16_t len)
+{
+	int rc;
+
+	if (len < 1) {
+		rc = generic_handler(peer, SOCP_WRITE_CGM_COMMUNICATION_INTERVAL,
+					SOCP_RSP_INVALID_OPERAND);
+		return rc;
+	}
+
+	if (req_operand[0] > 0 && req_operand[0] < BT_CGMS_COMM_INTERVAL_MIN) {
+		rc = generic_handler(peer, SOCP_WRITE_CGM_COMMUNICATION_INTERVAL,
+					SOCP_RSP_OUT_OF_RANGE);
+		return rc;
+	}
+
+	if (req_operand[0] == 0xFF) {
+		comm_interval = BT_CGMS_COMM_INTERVAL_MIN;
+	} else if (req_operand == 0) {
+		/* In this case, the server should stop notification.
+		 * Instead of cancel the work, we keep the work item
+		 * active, and inside the wotk handler we decide if
+		 * we should send notification. It makes sure the perioidic
+		 * task is submitted once in initialization function,
+		 * and avoids duplicate submission of the task.
+		 */
+		comm_interval = 0;
+	} else {
+		comm_interval = req_operand[0];
+	}
+
+	rc = generic_handler(peer, SOCP_WRITE_CGM_COMMUNICATION_INTERVAL,
+					SOCP_RSP_SUCCESS);
+	return rc;
+}
+
+static int read_cgm_communication_interval_handler(struct bt_conn *peer,
+			const uint8_t *req_operand, uint16_t len)
+{
+	uint8_t rsp[3];
+
+	rsp[0] = SOCP_READ_CGM_COMMUNICATION_INTERVAL_RESPONSE;
+	rsp[1] = comm_interval & 0xFF;
+	rsp[2] = (comm_interval >> 8) & 0xFF;
+
+	int rc;
+
+	rc = cgms_socp_send_response(peer, rsp, sizeof(rsp));
+	return rc;
+}
+
+int cgms_socp_recv_request(struct bt_conn *peer, const uint8_t *req_data, uint16_t req_len)
+{
+	int rc;
+
+	/* Early check of the data length. */
+	if (req_len == 0) {
+		return -ENODATA;
+	}
+
+	if (req_data[0] == SOCP_WRITE_CGM_COMMUNICATION_INTERVAL) {
+		rc = write_cgm_communication_interval_handler(peer,
+					req_len - 1 > 0 ? &req_data[1] : NULL,
```

Comment:
My idea is to do basic and common check in this entry point function, and let command-specific check reside in corresponding command handlers. It will make adding new command handler easier.